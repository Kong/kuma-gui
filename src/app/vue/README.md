---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

