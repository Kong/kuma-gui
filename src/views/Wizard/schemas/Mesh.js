module.exports = {
  // name: null,
  // type: 'Mesh', // or `kind` for Kubernetes
  mtls: {
    enabledBackend: null,
    backends: []
  },
  tracing: {
    defaultBackend: null,
    backends: [
      {
        name: null,
        type: null
      }
    ]
  },
  logging: {
    backends: [
      {
        name: null,
        format: '{ "destination": "%KUMA_DESTINATION_SERVICE%", "destinationAddress": "%UPSTREAM_LOCAL_ADDRESS%", "source": "%KUMA_SOURCE_SERVICE%", "sourceAddress": "%KUMA_SOURCE_ADDRESS%", "bytesReceived": "%BYTES_RECEIVED%", "bytesSent": "%BYTES_SENT%"}',
        type: null
      }
    ]
  },
  metrics: {
    enabledBackend: null,
    backends: [
      {
        name: null,
        type: null
      }
    ]
  }
}
