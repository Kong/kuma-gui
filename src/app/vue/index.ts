import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
