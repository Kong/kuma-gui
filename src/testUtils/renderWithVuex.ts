import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
