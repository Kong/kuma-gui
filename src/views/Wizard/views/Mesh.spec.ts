import userEvent from '@testing-library/user-event'
import { screen } from '@testing-library/vue'
import Mesh from './Mesh.vue'
import renderWithVuex from '@/testUtils/renderWithVuex'

describe('Mesh.vue', () => {
  const doStep = async (button: any, target: string | string[]) => {
    // Begin step

    expect(button).not.toHaveAttribute('disabled')

    // Check extended validation

    await userEvent.click(screen.getByText('Enabled'))

    expect(button).toHaveAttribute('disabled')

    if (typeof target === 'string') {
      await userEvent.type(screen.getByLabelText(target), 'fake-name')
    } else {
      const map = target.map(targetName => userEvent.type(screen.getByLabelText(targetName), 'fake-name'))

      await Promise.all(map)
    }

    expect(button).not.toHaveAttribute('disabled')

    await userEvent.click(screen.getByText('Next ›'))
  }

  it('passes whole wizzard and render yaml', async () => {
    const { container } = renderWithVuex(Mesh, {
      store: {
        modules: {
          config: { state: { tagline: 'Kuma', clientConfig: { environment: 'universal' } } },
        },
      },
      routes: [],
    })

    const nextButton = screen.getByText(/Next ›/i).closest('button')

    // basic validation on first step
    expect(nextButton).toHaveAttribute('disabled')

    await userEvent.type(screen.getByLabelText('Mesh name:'), 'default')

    // 1 step cd.

    await doStep(nextButton, 'Certificate name:')

    // 2 Step

    await doStep(nextButton, 'Backend name:')

    // 3 Step

    await doStep(nextButton, ['Backend name:', 'URL:'])

    // 4 Step

    await doStep(nextButton, 'Backend name:')

    await screen.findByText(/kumactl apply -f/)

    expect(container).toMatchSnapshot()
  })
})
