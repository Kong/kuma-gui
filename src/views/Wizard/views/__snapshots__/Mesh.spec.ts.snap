// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mesh.vue passes whole wizzard and render yaml 1`] = `
<div>
  <div
    class="wizard"
  >
    <div
      class="wizard__content"
    >
      <div
        class="wizard-steps"
      >
        <div
          class="wizard-steps__content-wrapper"
        >
          <header
            class="wizard-steps__indicator"
          >
            <ul
              aria-label="steptabs"
              class="wizard-steps__indicator__controls"
              role="tablist"
            >
              
              <li
                aria-controls="wizard-steps__content__item--0"
                aria-selected="false"
                class="is-complete wizard-steps__indicator__item"
              >
                <span>
                  General & Security
                </span>
              </li>
              <li
                aria-controls="wizard-steps__content__item--1"
                aria-selected="false"
                class="is-complete wizard-steps__indicator__item"
              >
                <span>
                  Logging
                </span>
              </li>
              <li
                aria-controls="wizard-steps__content__item--2"
                aria-selected="false"
                class="is-complete wizard-steps__indicator__item"
              >
                <span>
                  Tracing
                </span>
              </li>
              <li
                aria-controls="wizard-steps__content__item--3"
                aria-selected="false"
                class="is-complete wizard-steps__indicator__item"
              >
                <span>
                  Metrics
                </span>
              </li>
              <li
                aria-controls="wizard-steps__content__item--4"
                aria-selected="true"
                class="is-complete wizard-steps__indicator__item"
              >
                <span>
                  Install
                </span>
              </li>
              
            </ul>
          </header>
          <div
            class="wizard-steps__content"
          >
            <form
              autocomplete="off"
            >
              
              <div
                aria-labelledby="wizard-steps__content__item--0"
                class="wizard-steps__content__item"
                id="wizard-steps__content__item--0"
                role="tabpanel"
                tabindex="0"
              >
                <!--v-if-->
              </div>
              <div
                aria-labelledby="wizard-steps__content__item--1"
                class="wizard-steps__content__item"
                id="wizard-steps__content__item--1"
                role="tabpanel"
                tabindex="0"
              >
                <!--v-if-->
              </div>
              <div
                aria-labelledby="wizard-steps__content__item--2"
                class="wizard-steps__content__item"
                id="wizard-steps__content__item--2"
                role="tabpanel"
                tabindex="0"
              >
                <!--v-if-->
              </div>
              <div
                aria-labelledby="wizard-steps__content__item--3"
                class="wizard-steps__content__item"
                id="wizard-steps__content__item--3"
                role="tabpanel"
                tabindex="0"
              >
                <!--v-if-->
              </div>
              <div
                aria-labelledby="wizard-steps__content__item--4"
                class="wizard-steps__content__item"
                id="wizard-steps__content__item--4"
                role="tabpanel"
                tabindex="0"
              >
                
                <div>
                  <div>
                    <h3>
                       Install a new Mesh 
                    </h3>
                    <p>
                       Since the Kuma GUI is read-only mode to follow Ops best practices, please execute the following command in your shell to create the entity. Kuma will automatically detect when the new entity has been created. 
                    </p>
                    <div
                      class="tab-container"
                      data-testid="tab-container"
                    >
                      <!--v-if-->
                      <div
                        class="tab__content-container"
                      >
                        <div
                          class="k-tabs"
                          data-v-6bcd6661=""
                        >
                          <ul
                            aria-label="ktabs"
                            data-v-6bcd6661=""
                            role="tablist"
                          >
                            
                            <li
                              aria-controls="panel-0"
                              aria-selected="false"
                              class="tab-item"
                              data-v-6bcd6661=""
                              id="kubernetes-tab"
                              role="tab"
                              tabindex="0"
                            >
                              <a
                                class="tab-link"
                                data-v-6bcd6661=""
                              >
                                
                                Kubernetes
                                
                              </a>
                            </li>
                            <li
                              aria-controls="panel-1"
                              aria-selected="true"
                              class="active tab-item"
                              data-v-6bcd6661=""
                              id="universal-tab"
                              role="tab"
                              tabindex="0"
                            >
                              <a
                                class="tab-link"
                                data-v-6bcd6661=""
                              >
                                
                                Universal
                                
                              </a>
                            </li>
                            
                          </ul>
                          
                          <div
                            aria-labelledby="kubernetes-tab"
                            class="tab-container"
                            data-v-6bcd6661=""
                            id="panel-0"
                            role="tabpanel"
                            tabindex="0"
                          >
                            <!---->
                          </div>
                          <div
                            aria-labelledby="universal-tab"
                            class="tab-container"
                            data-v-6bcd6661=""
                            id="panel-1"
                            role="tabpanel"
                            tabindex="0"
                          >
                            
                            
                            <div
                              class="code-block"
                              tabindex="0"
                            >
                              <!--v-if-->
                              <div
                                class="code-block__content"
                              >
                                <!-- eslint-disable vue/no-v-html -->
                                <pre
                                  class="highlighted-code-block line-numbers language-bash"
                                  data-testid="highlighted-code-block"
                                  tabindex="0"
                                >
                                  <code
                                    class="language-bash"
                                  >
                                    <span
                                      class="token builtin class-name"
                                    >
                                      echo
                                    </span>
                                     
                                    <span
                                      class="token string"
                                    >
                                      "type: Mesh
name: default
mtls: 
  enabledBackend: fake-name
  backends: 
    - name: fake-name
      type: builtin
  enabled: true
tracing: 
  defaultBackend: fake-name
  backends: 
    - name: fake-name
      type: zipkin
      sampling: 99.9
      conf: 
        url: fake-name
logging: 
  backends: 
    - name: fake-name
      format: '{ start_time: ''%START_TIME%'', source: ''%KUMA_SOURCE_SERVICE%'', destination: ''%KUMA_DESTINATION_SERVICE%'', source_address: ''%KUMA_SOURCE_ADDRESS_WITHOUT_PORT%'', destination_address: ''%UPSTREAM_HOST%'', duration_millis: ''%DURATION%'', bytes_received: ''%BYTES_RECEIVED%'', bytes_sent: ''%BYTES_SENT%'' }'
      type: tcp
      conf: 
        address: '127.0.0.1:5000'
metrics: 
  enabledBackend: fake-name
  backends: 
    - name: fake-name
      type: prometheus
      conf: 
        port: 5670
        path: /metrics"
                                    </span>
                                     
                                    <span
                                      class="token operator"
                                    >
                                      |
                                    </span>
                                     kumactl apply 
                                    <span
                                      class="token parameter variable"
                                    >
                                      -f
                                    </span>
                                     -
                                    <span
                                      aria-hidden="true"
                                      class="line-numbers-rows"
                                    >
                                      <span
                                        data-line="1"
                                      />
                                      <span
                                        data-line="2"
                                      />
                                      <span
                                        data-line="3"
                                      />
                                      <span
                                        data-line="4"
                                      />
                                      <span
                                        data-line="5"
                                      />
                                      <span
                                        data-line="6"
                                      />
                                      <span
                                        data-line="7"
                                      />
                                      <span
                                        data-line="8"
                                      />
                                      <span
                                        data-line="9"
                                      />
                                      <span
                                        data-line="10"
                                      />
                                      <span
                                        data-line="11"
                                      />
                                      <span
                                        data-line="12"
                                      />
                                      <span
                                        data-line="13"
                                      />
                                      <span
                                        data-line="14"
                                      />
                                      <span
                                        data-line="15"
                                      />
                                      <span
                                        data-line="16"
                                      />
                                      <span
                                        data-line="17"
                                      />
                                      <span
                                        data-line="18"
                                      />
                                      <span
                                        data-line="19"
                                      />
                                      <span
                                        data-line="20"
                                      />
                                      <span
                                        data-line="21"
                                      />
                                      <span
                                        data-line="22"
                                      />
                                      <span
                                        data-line="23"
                                      />
                                      <span
                                        data-line="24"
                                      />
                                      <span
                                        data-line="25"
                                      />
                                      <span
                                        data-line="26"
                                      />
                                      <span
                                        data-line="27"
                                      />
                                      <span
                                        data-line="28"
                                      />
                                      <span
                                        data-line="29"
                                      />
                                      <span
                                        data-line="30"
                                      />
                                      <span
                                        data-line="31"
                                      />
                                    </span>
                                  </code>
                                </pre>
                                <!-- eslint-enable vue/no-v-html -->
                                
                                <div
                                  aria-controls="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                                  aria-expanded="false"
                                  data-v-9bbde9b2=""
                                  id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                                  role="button"
                                >
                                  
                                  <button
                                    class="code-block__copy-button"
                                    type="button"
                                  >
                                    <span
                                      class="kong-icon-copy kong-icon"
                                      data-v-637de6fa=""
                                    >
                                      <svg
                                        height="24"
                                        role="img"
                                        viewBox="0 0 32 32"
                                        width="24"
                                        xmlns="http://www.w3.org/2000/svg"
                                      >
                                        
  
                                        <title>
                                          Copy
                                        </title>
                                        
  
                                        <path
                                          d="M25.313 28v-18.688h-14.625v18.688h14.625zM25.313 6.688c1.438 0 2.688 1.188 2.688 2.625v18.688c0 1.438-1.25 2.688-2.688 2.688h-14.625c-1.438 0-2.688-1.25-2.688-2.688v-18.688c0-1.438 1.25-2.625 2.688-2.625h14.625zM21.313 1.313v2.688h-16v18.688h-2.625v-18.688c0-1.438 1.188-2.688 2.625-2.688h16z"
                                          fill="var(--blue-500)"
                                        />
                                        

                                      </svg>
                                      

                                    </span>
                                    <span
                                      class="kutil-sr-only"
                                    >
                                      Copy
                                    </span>
                                  </button>
                                  
                                  <transition-stub
                                    appear="false"
                                    css="true"
                                    data-v-9bbde9b2=""
                                    name="fade"
                                    persisted="false"
                                  >
                                    <div
                                      class="k-popover"
                                      data-v-9bbde9b2=""
                                      id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                                      role="region"
                                      style="width: 200px; max-width: 350px; max-height: auto; display: none;"
                                    >
                                      <!---->
                                      <div
                                        class="k-popover-content"
                                        data-v-9bbde9b2=""
                                      >
                                        
                                        <p>
                                          Code copied to clipboard
                                        </p>
                                        
                                      </div>
                                      <!---->
                                    </div>
                                  </transition-stub>
                                </div>
                                
                              </div>
                            </div>
                            
                            
                          </div>
                          
                        </div>
                        <!--v-if-->
                      </div>
                    </div>
                  </div>
                  <div
                    class="scanner"
                  >
                    <div
                      class="scanner-content"
                    >
                      <!-- loading state -->
                      <section
                        class="empty-state-wrapper"
                        data-v-5b55eb64=""
                      >
                        <div
                          class="empty-state-title"
                          data-v-5b55eb64=""
                        >
                          <!---->
                          <div
                            class="k-empty-state-title-header mt-4 mb-4"
                            data-v-5b55eb64=""
                          >
                            
                            <div
                              class="card-icon mb-3"
                            >
                              <span
                                class="kong-icon-spinner kong-icon"
                                data-v-637de6fa=""
                              >
                                <svg
                                  fill="rgba(0, 0, 0, 0.1)"
                                  height="42"
                                  role="img"
                                  viewBox="0 0 20 20"
                                  width="42"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  
  
                                  <title>
                                    Loading
                                  </title>
                                  
  
                                  <g>
                                    
    
                                    <path
                                      d="M5 10a5 5 0 1 0 5-5"
                                      fill="none"
                                      stroke="rgba(0, 0, 0, 0.1)"
                                      stroke-linecap="round"
                                      stroke-linejoin="round"
                                      stroke-width="2"
                                    />
                                    
  
                                  </g>
                                  

                                </svg>
                                

                              </span>
                            </div>
                            <!--v-if-->
                            
                            <h3>
                              Searching…
                            </h3>
                            
                            <!--v-if-->
                            
                          </div>
                        </div>
                        <div
                          class="empty-state-content"
                          data-v-5b55eb64=""
                        >
                          <div
                            class="k-empty-state-message mb-6"
                            data-v-5b55eb64=""
                          >
                            
                            
                            <p>
                              We are looking for your mesh.
                            </p>
                            
                            <!--v-if-->
                            
                          </div>
                          <div
                            class="k-empty-state-cta"
                            data-v-5b55eb64=""
                          >
                            
                            <!---->
                            
                          </div>
                        </div>
                      </section>
                    </div>
                  </div>
                </div>
                
              </div>
              
            </form>
          </div>
          <footer
            class="wizard-steps__footer"
          >
            <button
              class="medium rounded outline k-button"
              data-v-6bcbf05d=""
              style=""
              type="button"
            >
              
              <!---->
              
              
               ‹ Previous 
              
              <!---->
            </button>
            <button
              class="medium rounded primary k-button"
              data-v-6bcbf05d=""
              style="display: none;"
              type="button"
            >
              
              <!---->
              
              
               Next › 
              
              <!---->
            </button>
          </footer>
        </div>
        <aside
          class="wizard-steps__sidebar"
        >
          <div
            class="wizard-steps__sidebar__content"
          >
            
            <div
              class="wizard-steps__sidebar__item wizard-steps__sidebar__item--0"
            >
              
              <h3>
                Mesh
              </h3>
              <p>
                 In Kuma, a Mesh resource allows you to define an isolated environment for your data-planes and policies. It's isolated because the mTLS CA you choose can be different from the one configured for our Meshes. Ideally, you will have either a large Mesh with all the workloads, or one Mesh per application for better isolation. 
              </p>
              <p>
                <a
                  href="https://kuma.io/docs/1.2.0/policies/mesh/?utm_source=Kuma&utm_medium=Kuma-GUI"
                  target="_blank"
                >
                   Learn More 
                </a>
              </p>
              
            </div>
            <div
              class="wizard-steps__sidebar__item wizard-steps__sidebar__item--1"
            >
              
              <h3>
                Did You Know?
              </h3>
              <p>
                 As you know, the GUI is read-only, but it will be providing instructions to create a new Mesh and verify everything worked well. 
              </p>
              
            </div>
            
          </div>
        </aside>
      </div>
    </div>
  </div>
</div>
`;
